Knife SSH prefix option

mal opened the pull request:
### Description

```sh
# existing usage (no change)
$ knife ssh 'name:app-server' 'echo hello' 
ec2-10-0-0-1.compute-1.amazonaws.com hello

# new functionality
$ knife ssh --prefix-attribute name 'name:app-server' 'echo hello'
app-server hello
```

Allows an attribute to be specified which will determine how a node is identified in the output of `knife ssh` commands. It uses the long flag `--prefix-attribute` ~~and the short flag `-b`. The short flag was picked as a placeholder, chosen because `-p` is already in use for `--port`, this can be changed prior to merging to align with any sort of option/flag guidelines that may exist.~~ The config setting `Chef::Config[:knife][:prefix_attribute]` is also supported.

This feature makes use of the session properties assignment functionality documented [here](https://github.com/jamis/net-ssh-multi/blob/c0b0a0b9a84d0e2ba58dd6f78e0d5e28900e18c4/lib/net/ssh/multi/session.rb#L320) to attach the prefix to the session, and renames some internal keys (`config`, `knife_config`) in order to aid understanding of the code when reading and provide clear distinctions between the attribute keys used for the SSH target and the prefix.

### Issues Resolved

Resolves #6573

---

~~Tests seem to have been broken before I got here 🙁~~ Looks like it's been fixed \o/

mal commented:
Test failures appear to be related to file permissions in Travis, and were already broken in appveyor rather than an of the code changes here :/

mal commented:
Tests seem pretty broken before I got here 🙁

mal commented:
Rebased on master to get passing tests \o/

lamont-granquist commented on lib/chef/knife/ssh.rb:
```diff
@@ -59,6 +59,12 @@ class Ssh < Knife
+        :short => "-b ATTR",
+        :long => "--prefix-attribute ATTR",
+        :description => "The attribute to use for prefixing the ouput - default depends on the context",
+        :proc => Proc.new { |key| Chef::Config[:knife][:prefix_attribute] = key.strip }
```
I'm reasonably certain that if you name your option `:prefix_attribute` and delete this proc that it merges correctly now (and has for a very long time, but we have never cleaned up this from the knife attributes), and in particular you should make sure that the Chef::Config option and the Mixlib::Cli option here match.

lamont-granquist commented on lib/chef/knife/ssh.rb:
```diff
@@ -181,18 +187,34 @@ def configure_session
+        if node["prefix_cli"]
+          Chef::Log.debug("Using node attribute from CLI '#{config[:prefix]}' as the prefix: #{node["prefix_cli"]}")
+          node["prefix_cli"]
+        elsif Chef::Config[:knife][:prefix_attribute]
```
Also here you should use `config[:prefix_attribute]` to get the merging magic.

lamont-granquist commented on lib/chef/knife/ssh.rb:
```diff
@@ -212,14 +234,24 @@ def search_nodes
+        # if we've configured a prefix attribute in our config
+        if Chef::Config[:knife][:prefix_attribute]
+          required_attributes[:prefix_config] = Chef::Config[:knife][:prefix_attribute].split(separator)
+        end
```
again this should just be `config[:prefix_attribute]` and let the knife cli/config merging code handle it.

lamont-granquist commented on lib/chef/knife/ssh.rb:
```diff
@@ -181,18 +187,34 @@ def configure_session
-          Chef::Log.debug("Using node attribute #{Chef::Config[:knife][:ssh_attribute]}: #{node["knife_config"]}")
-          node["knife_config"]
+          Chef::Log.debug("Using node attribute from config '#{Chef::Config[:knife][:ssh_attribute]}' as the ssh target: #{node["ssh_config"]}")
+          node["ssh_config"]
```
this looks like it changed an external API on the node?

mal replied:
`node` in this context is the response from the custom `query.search` call we make in `search_nodes`.

coderanger commented:
This doesn't seem common enough to use a short option too, can we just do the long option?

mal commented:
All the changes here are based on mirroring the current approach taken with the ssh attribute option. I'm not at all familiar with the "merging magic" stuff mentioned, so maybe that's a module level refactor worth doing as a separate body of work?

Dropping the short flag seems like a good idea, I'll sort that out in a bit 👍

lamont-granquist commented:
yes, the existing code may just be broken, and may be more complicated than it needs to be, please don't copy that.

if you just name the cli option and the knife config option the same and access them through the `config` hash it works correctly and the `Proc` in the cli option can be dropped.  anything else and you're copypasta'ing dubious code around.

mal commented:
- Sorted out the newer config merge magic for both `:ssh_attribute` and `:prefix_attribute`
- Renamed `node` in the `get_*_attribute` methods so it's clear that they're not operating on real nodes
- Removed the short flag in favour of just having the long flag

mal commented:
Flaky tests are back, but not related to this PR 🙁

lamont-granquist commented:
coolness, merging...
